src/languages/haskell.js

```diff
@@ -154,10 +154,16 @@ export default function(hljs) {
       PREPROCESSOR,
 
       // Literals and names.
+      hljs.inherit(hljs.BINARY_NUMBER_MODE, {
+        begin: '\\b(0b[01_]+)'
+      }),
+
+      hljs.inherit(hljs.C_NUMBER_MODE, {
+        begin: '(-?)(\\b0[xX][a-fA-F0-9_]+|(\\b(\\d|_)+(\\.(\\d|_)*)?|\\.(\\d|_)+)([eE][-+]?(\\d|_)+)?)'
```

Comment:
Generally we do not match `-` since it could often by unary or binary and we do not try and figure out such distinctions. (very hard to do also without look-behind)

Also please add additional tests for some of the `e` variants, etc.

Reply:
Are there two variants here? (hex and regular?)  It seems so.  Please split them out into `variants` for readability and maintenance rather than using a more complex single regex.  See many other grammars for examples.

Reply:
Ah right, I just copied it from `C_NUMBER_MODE`. I'll make a few tests though :-)

Reply:
Well, if it's exactly the same it of course shouldn't need to be copied at all... but if it's now specific to this language then we should clean it up into variants for future use - which is just the general policy on such things.

I'd suggest we could perhaps clean up `C_NUMBER_MODE` as well but too many grammars are likely dependent on it being *exactly* how it is now - but that doesn't mean we can't have nicer and easy to read matches in the individual grammars. :)

Reply:
I'll add such thoughts to the v12 list.  v11 is already big enough.

Reply:
Ah let me clarify a bit: I copied it and then added `|_` to optionally match the underscores. 

I'm a bit fuzzy on what you want me to do; should I just make a clean version for Haskell numbers?

Reply:
See https://github.com/highlightjs/highlight.js/blob/main/src/languages/swift.js#L121 for an example.

When we start adding *custom* numeric rules to grammars then we always do it as nicely as possible (for future readability and maintenance).  That means separate `variants` for hex, binary, etc... (as you see in Swift) not just one large regex with multiple rules "hidden" inside.

Reply:
Excellent. I'll get to it tonight. Thanks for the quick responses :)